### ISSUE TITLE
Panic: "internal compiler error: ident only path should have been covered already"

### ISSUE THREAD

**Issue by insaneinside** [rust-lang/rust#22426]
This error is triggered by providing type parameters to a generic type in a `match` pattern.

e.g.
```rust
#![crate_type = "staticlib"]

pub struct Foo<T>(T, T);

impl<T> Foo<T> {
    fn foo(&self) {
        match *self {
            Foo<T>(x, y) => println!("Goodbye, World!")
        }
    }
}
```
[Playpen here.](http://is.gd/qUaQmI)

**Comment by mzap** [comment]
This is not the only way to make it panic.

http://is.gd/LcO1Pw
Using nightly.

> rustc 1.0.0-nightly (b63cee4a1 2015-02-14 17:01:11 +0000)
> binary: rustc
> commit-hash: b63cee4a11fcfecf20ed8419bc3bd6859e6496bc
> commit-date: 2015-02-14 17:01:11 +0000
> host: x86_64-unknown-linux-gnu
> release: 1.0.0-nightly

**Comment by Undeterminant** [comment]
I also have this problem when pattern matching types with `<>`.

**Comment by Kimundi** [comment]
It seems I introduced this bug with [#22158](https://github.com/rust-lang/rust/pull/22158/files#diff-da9d34ca1d0f6beee2838cf02e07345cL3535), sorry :frowning:. 

The ICE happens because I had refactored parts of the parser that seemed to be dead code, apparently I was wrong about that though, and neither testsuite nor the rest of the compiler codebase walked into it.

The good news is that its straightforward to fix, someone just needs to revert the linked part of the diff (And add a test for it...).

**Comment by bombless** [comment]
Seems that I made a different approach.. did I do something wrong? :P

**Comment by Kimundi** [comment]
Was it legal before to have type paramters there, or did it lead to an error in a later stage?

**Comment by bombless** [comment]
It is legal, see my comment here https://github.com/rust-lang/rust/issues/22546#issuecomment-75426130

**Comment by bombless** [comment]
I made a patch for this, change `LifetimeAndTypesWithColons` to `LifetimeAndTypesWithColons(bool)` to disable type parameters in path in some condition to reject type parameters in pattern, and it is easy.

**Comment by Kimundi** [comment]
Okay, so allowing `Foo<T>` is wrong, because `Foo::<T>` should be accepted?

**Comment by bombless** [comment]
` Foo::<T> ` make sense to me, and I'd admit we don't really need it in pattern.
But there's another thing a little bit related and I think we can do some effort for it: https://github.com/rust-lang/rust/issues/22644

**Comment by Kimundi** [comment]
Well, thats an unrelated issue to this one.

Looking at your patch, it might be possible to catch the `Foo<T>` case by removing the `token::Lt` in the earlier `can_be_enum_or_struct` initialization, since `IDENT <` is apparently not legal in a pattern: No comparison operators, and generics are parsed as `Foo::<T>` already.

**Comment by bombless** [comment]
:smile: Yeah I have the same feeling.
But we can also catch the error just before anything go wrong, it may even save some CPU cycles on early error handling (partly joking).

**Comment by Kimundi** [comment]
Right, detecting the misplaced `<` and giving a proper error message that guides one to use the `Foo::<T>` form would be the best.

**Comment by bombless** [comment]
>Right, detecting the misplaced `<` and giving a proper error message that guides one to use the  `Foo::<T>`  form would be the best.

Agreed.
Do you think we may wait for a `>` for that error message?
I think it would be great if we can directly point out `Foo<T>` should be `Foo::<T>`.
As it is a status we are doing error handling, we cannot make mess that much.
And if we meet something that don't even look like a path, then we just stop reading tokens and point out token `<` is legal.

This kind of code may be hard to maintain, though.
